### PR TITLE
[terra-form-select] Max select count

### LINKED ISSUE #2227: [terra-form-select] Ability to limit the maximum number of selections for multi-select
# Feature Request

## Description
The `tag` and `multiple` select variants should provide the ability to limit the maximum number of selected items. Currently, users can select all enabled options in the menu.

### PR DESCRIPTION
**Pull request by benbcai**
### Summary
Add option to limit the maximum number of selections in a multi-select.

### Additional Details
Closes #2227.
- Added new prop `maxSelectionCount` to SelectField. It was decided to only expose this prop to SelectField and not Select because the help text in SelectField is utilized to show a show a text to indicate that the maximum number of selection has reached. This text will be prepended to the passed in help text if one is provided.
- Applicable only to the `tag` or `multiple` variants
- When the maximum number of selection is reached, all other unselected options will be disabled.
- Strings still need translations.

### Example Deployment Links:
https://terra-core-deployed-pr-2244.herokuapp.com/#/components/terra-form-select/form-select/select-field

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

**Comment by bjankord** [comment]
@benbcai This looks good, one think that I saw needs added is the ability to deselect an option with a keyboard. I think this can be added later if we want to keep the scope as-is or include now.

**Comment by benbcai** [comment]
@bjankord The change to deselect an option with a keyboard was already made by @JakeLaCombe in #2217. That change is now carried over in #2236. Here is the exact change: https://github.com/cerner/terra-core/pull/2236/files#diff-a7fffd10160b31b692e4d703e8ff5280R303

**Comment by bjankord** [comment]
@benbcai Sounds good, let's get @mmalaker to functionally verify and @neilpfeiffer to UX verify and then this is good to merge.

File: packages/terra-form-select/src/SelectField.jsx [hunk 1]
```diff
@@ -112,60 +119,111 @@ const defaultProps = {
...
+
+    this.state = { help: undefined };
+    this.handleChange = this.handleChange.bind(this);
+  }
+
+  handleChange(value) {
+    const { maxSelectionCount, variant } = this.props;
+
+    if (maxSelectionCount !== undefined && MenuUtil.allowsMultipleSelections(variant) && Array.isArray(value) && value.length >= maxSelectionCount) {
+      let helpWithMaxSelectionCount = this.context.intl.formatMessage({ id: 'Terra.form.select.maxSelectionHelp' }, { text: maxSelectionCount });
```
Comment (neilpfeiffer) [inline packages/terra-form-select/src/SelectField.jsx]: pretty sure we want the `helpWithMaxSelectionCount` to be injected always, not conditional if maxSelectionCount is reached.

File: packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx [hunk 2]
```diff
@@ -26,6 +28,11 @@ const DocPage = () => (
       title: 'Select Field Example',
       example: <SelectFieldExample />,
       source: SelectFieldExampleSrc,
+    }, {
+      title: 'Maximum Selection',
```
Comment (neilpfeiffer) [inline packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx]:
```suggestion
      title: 'Maximum Selection (applies to only \'multiple\' and \'tag\')',
```

 title: 'Clear', -> ' Ability to Clear Selection (applies to only 'default', 'combobox', and 'search')',

File: packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx [hunk 3]
```diff
@@ -26,6 +28,11 @@ const DocPage = () => (
...
+      description: 'A multiple variant that allows up to 3 selections',
```
Comment (neilpfeiffer) [inline packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx]:
```suggestion
      description: 'A multiple selection variant that allows up to a set number of selections. (Not applicable to  \'default\', \'combobox\', and \'search\' variants) ',
```


(applies to only)',
      description: 'An example with an option to clear the selected item. Not applicable to \'multiple\' or \'tag\' variants.',

**Comment by mmalaker** [comment]
Functional verification completed. The max select example functions as expected. +1 UX changes.

File: packages/terra-form-select/src/SelectField.jsx (hunk 1 above)
Comment (benbcai) [reply packages/terra-form-select/src/SelectField.jsx]: Ok. I'm updating it to always show the "Limit x items" in the help text.

File: packages/terra-form-select/CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,8 @@ ChangeLog
 
 Unreleased
 ----------
+### Added
+* Added option to limit the maximum number os selection count for multi-select variants.
```
Comment (StephenEsser) [inline packages/terra-form-select/CHANGELOG.md]: Small typo, `os`

Comment (benbcai) [reply packages/terra-form-select/CHANGELOG.md]: Fixed typo: https://github.com/cerner/terra-core/pull/2244/commits/4eda686ca929d32078c2db76c42660c4e94a7cd6

File: packages/terra-form-select/src/SelectField.jsx (hunk 1 above)
Comment (benbcai) [reply packages/terra-form-select/src/SelectField.jsx]: Updated: https://github.com/cerner/terra-core/pull/2244/commits/3c8375b9ae7e72f9eb55a40a15eab66819461ff7

File: packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx (hunk 2 above)
Comment (benbcai) [reply packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx]: Updated the example title: https://github.com/cerner/terra-core/pull/2244/commits/3c8375b9ae7e72f9eb55a40a15eab66819461ff7

File: packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx (hunk 3 above)
Comment (benbcai) [reply packages/terra-form-select/src/terra-dev-site/doc/form-select/SelectField.2.doc.jsx]: Updated the example description: https://github.com/cerner/terra-core/pull/2244/commits/3c8375b9ae7e72f9eb55a40a15eab66819461ff7

**Comment by neilpfeiffer** [comment]
Tested: PR #2244
 
**Reviewed:**
- Resolves adding max-selection count to `terra-form-select/lib/SelectField`
 
 
Verified by @neilpfeiffer, ready for release.
 
**Pass** |
-----------|
UX Review|

**Pull request merged**